### PR TITLE
Promote variables in configure (POEM_003)

### PR DESCRIPTION
**Pull request by DKilkenny**
https://www.pivotaltracker.com/story/show/170749130

### CODE REVIEW

File: openmdao/core/tests/test_group.py [hunk 1]
```diff
@@ -226,6 +318,42 @@ def test_group_rename_connect(self):
...
+
+            def configure(self):
+                self.promotes('comp1', any=['*'])
+
+        top = om.Problem(model=SimpleGroup())
+        top.setup()
+
+        self.assertEqual(top['x'], 5)
+
+    def test_promotes_inputs_and_outputs(self):
```
Comment (Kenneth-T-Moore) [inline openmdao/core/tests/test_group.py]: Tests that are used in the feature docs need to include all imports.  If you forget to do this, it should fail when the docs build (tests are run independently in that process), so if you built these docs, something isn't working right there.

File: openmdao/core/group.py [hunk 2]
```diff
@@ -1374,6 +1374,34 @@ def _setup_transfers(self, recurse=True):
...
+        inputs : Sequence of str or tuple
+            A Sequence of input names (or tuples) to be promoted. Tuples are
+            used for the "promote as" capability.
+        outputs : Sequence of str or tuple
+            A Sequence of output names (or tuples) to be promoted. Tuples are
+            used for the "promote as" capability.
+        """
+        subsys = getattr(self, subsys_name)
+        if any:
+            subsys._var_promotes['any'] = any
```
Comment (Kenneth-T-Moore) [inline openmdao/core/group.py]: This looks like it would replace any existing values in the _var_promotes dictionary.  I think it is more correct to append them to the existing dictionary instead, in case the user want to call `promotes` twice on the same subsystem.

File: openmdao/core/tests/test_group.py (hunk 1 above)
Comment (DKilkenny) [reply openmdao/core/tests/test_group.py]: Yeah I forgot. This was added in the latest commit

File: openmdao/core/group.py (hunk 2 above)
Comment (DKilkenny) [reply openmdao/core/group.py]: Fixed

File: openmdao/core/tests/test_group.py [hunk 3]
```diff
@@ -96,6 +96,123 @@ def configure(self):
...
+        top.setup()
+
+        self.assertEqual(top['x'], 5)
+        with self.assertRaises(UnboundLocalError) as cm:
+            top['a']
+
+        self.assertEqual(str(cm.exception),
+                         "local variable 'val' referenced before assignment")
+
+    def test_multiple_promotes(self):
```
Comment (naylor-b) [inline openmdao/core/tests/test_group.py]: I don't see any tests in here that do promotes with aliasing.  We should probably have a couple of those, one with multiple nested groups that does promotion with aliasing successfully and another that maybe tries and fails to promote an aliased child name using the original name.

Comment (DKilkenny) [reply openmdao/core/tests/test_group.py]: Added

File: openmdao/core/tests/test_group.py [hunk 4]
```diff
@@ -96,6 +96,146 @@ def configure(self):
...
+                self.promotes('comp2', outputs=['b'])
+
+        top = om.Problem(model=SimpleGroup())
+        top.setup()
+
+        self.assertEqual(top['b'], 1)
+        with self.assertRaises(UnboundLocalError) as cm:
+            top['a']
+
+        self.assertEqual(str(cm.exception),
```
Comment (naylor-b) [inline openmdao/core/tests/test_group.py]: This error message seems unrelated to the action being checked.  `top['a']` results in `local variable "val" referenced before assignment'?  Something doesn't seem right here.  This happens in several places.

File: openmdao/core/tests/test_group.py [hunk 5]
```diff
@@ -96,6 +96,146 @@ def configure(self):
...
+                         "local variable 'val' referenced before assignment")
+
+    def test_promotes_inputs_in_config(self):
```
Comment (naylor-b) [inline openmdao/core/tests/test_group.py]: I think we still need a test where a subsystem promotes a variable to one alias, and then a parent tries to promote it to a different alias.  That should raise an exception.  Promoting it to the same alias would be ok.

File: openmdao/core/system.py [hunk 6]
```diff
@@ -2038,9 +2038,27 @@ def resolve(to_match, io_types, matches, proms):
...
                 else:
                     call = 'promotes_%ss' % io_types[0]
-                raise RuntimeError("%s: '%s' failed to find any matches for the following "
-                                   "names or patterns: %s.%s" %
-                                   (self.msginfo, call, sorted(not_found), empty_group_msg))
+
+                for p in patterns:
+                    for name, alias in renames.items():
+                        if fnmatchcase(name, p):
+                            raise RuntimeError("%s: %s '%s' matched '%s'. You are attempting to "
```
Comment (naylor-b) [inline openmdao/core/system.py]: Since we have the information available here, we should maybe say " ... matched 'bb' but 'bb' was aliased to 'xx'."

File: openmdao/core/tests/test_group.py [hunk 7]
```diff
@@ -96,6 +96,223 @@ def configure(self):
...
+
+            def configure(self):
+                self.sub.promotes('comp1', inputs=['bb'])
+
+        top = om.Problem(model=TopGroup())
+
+        # with self.assertRaises(RuntimeError) as context:
+        top.setup()
+
+        # self.assertEqual(str(context.exception),
```
Comment (naylor-b) [inline openmdao/core/tests/test_group.py]: Get rid of commented out code.

File: openmdao/core/system.py (hunk 6 above)
Comment (DKilkenny) [reply openmdao/core/system.py]: fixed

File: openmdao/core/group.py [hunk 8]
```diff
@@ -1374,6 +1374,43 @@ def _setup_transfers(self, recurse=True):
...
+        if any:
+            subsys._var_promotes['any'].extend(any)
+        if inputs:
+            subsys._var_promotes['input'].extend(inputs)
+        if outputs:
+            subsys._var_promotes['output'].extend(outputs)
+
+        list_comp = [i if isinstance(i, tuple) else (i, i) for i in subsys._var_promotes['input']]
+
+        # Fix this
```
[the remark this answers is not in the document]
Comment (DKilkenny) [reply openmdao/core/group.py]: fixed

File: openmdao/core/tests/test_group.py [hunk 9]
```diff
@@ -985,8 +980,7 @@ def test_promote_not_found1(self):
         with self.assertRaises(Exception) as context:
             p.setup()
         self.assertEqual(str(context.exception),
-                         "ExecComp (C2): 'promotes_outputs' failed to find any matches for "
-                         "the following name: ['x*'].")
+                         "ExecComp (C2): 'promotes_outputs' failed to find any matches for the following pattern: x*.")
```
Comment (naylor-b) [inline openmdao/core/tests/test_group.py]: should put pattern in quotes

Comment (DKilkenny) [reply openmdao/core/tests/test_group.py]: fixed
